api/proto/config/devconfig.proto

```diff
@@ -36,19 +36,38 @@ message EdgeDevConfig {
         DeviceOpsCmd reboot = 9;
         DeviceOpsCmd backup = 10;
 
+        // XXX - Should we transition these over to use Maps as well?
         repeated ConfigItem configItems = 11;
+        // XXX - Should we transition these over to use Maps as well?
         repeated SystemAdapter systemAdapterList = 12;
+        // XXX - Should we transition these over to use Maps as well?
         repeated PhysicalIO deviceIoList = 13;
 
         // Override dmidecode info if set
         string manufacturer = 14;
         string productName = 15;
 
+        // XXX - Should we transition these over to use Maps as well?
         repeated NetworkInstanceConfig networkInstances = 16;
 
         // Information saved by device to make it easier to find in the controller
         string enterprise = 17;
         string name = 18;
+
+        // XXX - the following are needed ONLY if we want to move the Arrays
+        //  to maps. Advantage with maps is that we automatically get uniqueness
+        //  for primary keys. They also become easy to validate.
+        // newConfigItems
+        map <string, ConfigItem> configItemsMap = 20;
+        map <string, NetworkInstanceConfig> newNetworkInstancesMap = 22;
+
+        // deviceIoListMap - Map of phydical Adapters present on the device.
+        //  These attributes generally don't change.
+        map <string, PhysicalIO> deviceIoListMap = 23;
+        // systemAdapterMap - Configuration of the IoAdapters
+        map <string, SystemAdapter> systemAdapterMap = 24;
+
```

Comment:
Makes sense. I like this. Does this mean that the device need not do any reference checks?
eg:- system adapter pointing to a network and the network is not present in configuration.

Even if the device needs to do such checks, it will be easy with these changes.
Should we also have an API like EdgeDevStatus that the device sends back to cloud, telling if the configuration was accepted/rejected with corresponding error/success message string and time of acceptance/reject? This could purely be a zedagent level acceptance status (not wait for other agents). For example, zedagent finds configuration for some new feature like a new type for system adapter that we do not support and wants to reject (not sure if cloud already takes care of such checking).

This new thing can be done as part of a different PR also. Just putting forward my 2 cents for now.

Reply:
Yes.. we should do it separately. When we add this ability to error check and reject entire config, I would add this as a first class field in ZInfoDevice.